### ISSUE TITLE
Instructors need to understand details of role mapping for each tool

### ISSUE THREAD

**Issue by mccahill** [DukeLearningInnovation/kits#115]
As an instructor or instructor's delegate I need to understand how the community roles defined in the this application will map to roles/capabilities in the tools I am using. Because we cannot generally change the terminology or capabilities of the tools the application links to, there may not be the same number of roles possible and the roles may have different (or overlapping) names. We commonly lump groups together (i.e. treat course auditors and visitors  as the same group when provisioning them into Wordpress), but the instructor needs to understand what this mapping is, and that the mapping may be different for each external tool they choose. 

In the case of toolkits 1.0, the toolkits roles include: 
- Instructor      
- Manager         
- TA              
- Developer       
- Mentor          
- Student         
- Auditor         
- Visitor   
- Admin
- Contributor

**Comment by profmikegreene** [comment]
Going to contradict a previous comment https://github.com/DukeLearningInnovation/learn.duke/issues/93#issuecomment-383646570. 

What if we took the Slack approach and leaders could only add people as member or leader? 

Many teaching and learning tools come out of the box with teacher and student roles and that's it. This strategy would make it easier to map to more things and easier to communicate roles and how they map.

I wonder if a better response to users saying 'I need more variety in roles' is to move forward on custom roles instead of providing more by default.

I'm trying to get some data on how folks use Sakai to see what patterns emerge that would validate or contradict this hypothesis. Thoughts?

**Comment by LASHirsh** [comment]
I'm definitely open to this approach and am in favor of fewer roles overall than in Toolkits 1.0. 

Still thinking about four though:

- Manager
- Assistant
- Participant
- Visitor

Maybe it does not need to be the prerogative of this new application to do complex mapping or explain the various roles of and mapping to other applications at all, especially since one of our objectives is to be able to integrate many more tools in the future without a lot of work and making this unnecessarily complicated/confusing for end users. Community managers would use Toolkits to do very basic provisioning: create an instance of the external tool and give members some type of access. Setting up any new tool may mean some degree of updating permissions for the roster within that new tool. I don't think it's an unrealistic expectation but could be wrong.  

We need to be clear what "custom roles" are though. I don't think everyone on the team understands this the same way, but it has a lot of implications. Can you explain a bit more what you mean?

**Comment by JustinJohnsen** [comment]
I feel pretty strongly that the only default options should be roles we expect to be filled in every community (or 99% of them). To me, that means just instructor and student, if we are focused on campus classes.

You all know I'm an advocate for the custom roles option. I have described some already about being able to customize a role's access to different tools that are part of the community and also manage their ability to edit the community itself. This is where I think we'd provide some light example of how the customization works, like adding a class visitor, or guest lecturer, or TA, or staff assistant. Each of those may occur in many classes, but in each class, the instructor may have very specific ideas about what they want a person in that role to be able to do. I don't see how we can effectively provide the right kind of access to make everyone happy out of the box, so I feel that customization here is crucial and adds tremendous flexibility while not adding much confusion per class. However, I would want input from developers about how much work this would be to manage on the back end.

As a simplified example for discussion purposes, let's say there is just an admin/member setting for each tool and for the community editing. Perhaps an instructor chooses two tools to add to her community, PlayPosit and WarpWire. The default instructor role would have 'admin' access set across both the community edit option and the two tools. The default student role would have 'member' access set across those three. An instructor might have a TA, in which case they may want the TA to be able to create and grade PlayPosit activities, but not be admin in the community or in WarpWire, so they create a custom role (which they could label TA, Teaching Assistant, Grader, etc...) and the TA gets 'admin' rights in PlayPosit, but 'member' rights in the community edit option and in WarpWire. 

Further along, the instructor needs a staff assistant to handle the add/drop deadline, so they make a new custom role called Staff Assistant (or whatever they want) and assign admin rights to community editing, but member rights only to PlayPosit and WarpWire. Later still, the instructor has a guest lecturer join the class for a couple sessions, and wants to give them the ability to provide some flipped classroom videos, so they create a custom role Guest Lecturer and assign admin rights for WarpWire, but member rights for the community editing and for PlayPosit.

In each of the custom situations above, I would definitely not want to assign one of the default roles (Instructor/Student) to any of these new users. And I would only want to give admin rights in specific tools, rather than across the board, in many cases. In fact, there may also be a need for a third 'per tool' setting of 'none' in access rights, if even 'member' rights would be undesirable.

Current Toolkits already assigns these 'per tool' access rights, they just have a fixed combination of them for each Toolkits role. I'm suggesting that we keep that function, but make it more flexible in use. I believe each instructor of a community would only need 3-4 of these roles, which makes things look simpler to them inside their own community, but that across all communities, the number of possible combinations of tools and access rights could be in the thousands (10 available tools and 2 access rights settings already offers at least 1,000 different possible combinations).

**Comment by profmikegreene** [comment]
![tableau_-_sakai-roles](https://user-images.githubusercontent.com/3632822/39593633-5d6c24f6-4ed8-11e8-901d-2b7ba85497e0.png)

This is % of Total Users  by role in all Sakai sites created in any given year. Indicates to me that having three roles Leader, Member, and Read-only would account for the majority of use cases. I'm trying to generate the same viz by number of sites instead of total users to see if it looks differently

**Comment by profmikegreene** [comment]
I updated your comment to get the table formatting @JustinJohnsen.

There's definitely something to explore here for MVP2 that would be good for the dev team to plan for if possible. 

@ewendland23  @thegreenbot @noahhh what does it mean for you if we move forward in MVP1 with 2 community roles (whatever we call leader and member) + the few system roles (admin, superadmin, content manager, etc) with the expectation that MVP2 will include some type of community custom role and/or permissioning system that leaders can manage?


---


![tableau_-_sakai-roles-2](https://user-images.githubusercontent.com/3632822/39594987-2eeb8532-4edc-11e8-8268-30a53535c5bb.png)

This is % of Sakai Sites that have at least 1 user of each role created in any given year. Course sites in the left column, project in the right. I actually thought the TA/Course Builder numbers were going to be higher
Tableau workbook available https://tableau.oit.duke.edu/#/site/prod/workbooks/11640/views
Query to generate data
```sql
with roles as (
    select
    site_id
    ,user_role
    ,count(user_role) as users
    FROM duke_logs.sakai_site_user
    group by site_id, user_role
), total_users as (
    select 
    site_id
    ,count(user_id) as total_users
    from duke_logs.sakai_site_user
    group by site_id
), total_sites as (
    select
    type
    ,datepart('year', createdon) as year
    ,count(site_id) as total_sites
    from duke_logs.sakai_site
    group by type, year
)
SELECT
sakai_site.site_id
,user_role
,sum(users) as users
,max(total_users) as total_users
,max(total_sites) as total_sites
,max(sakai_site.type) as type
,max(createdon) as createdon
,max(modifiedon) as modifiedon
FROM duke_logs.sakai_site
join roles on sakai_site.site_id = roles.site_id
join total_users on sakai_site.site_id = total_users.site_id
join total_sites on date_part('year', sakai_site.createdon) = total_sites.year and sakai_site.type = total_sites.type
group by 1,2
order by 1,2
```

**Comment by swwalker** [comment]
closed - referenced in #221
